internal/metadata/metadata.go

```diff
@@ -72,3 +75,46 @@ func Set(addr resolver.Address, md metadata.MD) resolver.Address {
 	addr.Attributes = addr.Attributes.WithValue(mdKey, mdValue(md))
 	return addr
 }
+
+// Validate returns an error if the input md contains invalid keys or values.
+//
+// if the header is not pseudo-header, there are check items:
+// - header names contain one or more characters from this set [0-9 a-z _ - .]
+// - if the header-name ends with a "-bin" suffix, the header-value could contain an arbitrary octet sequence. So no real validation required here.
+// - if header-name does not end with a "-bin" suffix, header-value should only contain one or more characters from the set ( %x20-%x7E ) which includes space and printable ASCII.
+func Validate(md metadata.MD) error {
+	for k, vals := range md {
+		// pseudo-header will be ignored
+		if k[0] == ':' {
+			continue
+		}
+		// check key, for i that saving a conversion if not using for range
```

Comment:
I don't understand this.  Why would `for _, r := range(k) {` be worse?

Reply:
Range is different for strings (runes vs bytes etc): go.dev/play/p/2B-0HiWektw
https://github.com/grpc/grpc-go/pull/4886#discussion_r760434233

Reply:
IIUC all of these will work correctly anyway since we don't allow anything besides low ASCII, right?